pkg/liqo-controller-manager/webhooks/shadowpod/webhook.go

```diff
@@ -0,0 +1,301 @@
+func (spv *Validator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	webhooklog.Info(fmt.Sprintf("Operation: %s", req.Operation))
+
+	if !spv.peeringCache.ready {
+		webhooklog.Info("PeeringCache not ready")
+		return admission.Errored(http.StatusInternalServerError, fmt.Errorf("initialization in progress"))
+	}
+
+	switch req.Operation {
+	case admissionv1.Create:
+		return spv.HandleCreate(ctx, &req)
+	case admissionv1.Delete:
+		return spv.HandleDelete(ctx, &req)
+	case admissionv1.Update:
+		return spv.HandleUpdate(ctx, &req)
+	default:
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("unsupported operation %s", req.Operation))
+	}
+}
+
+// HandleCreate is the function in charge of handling Creation requests.
+func (spv *Validator) HandleCreate(ctx context.Context, req *admission.Request) admission.Response {
+	shadowpod, err := spv.DecodeShadowPod(req.Object)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, fmt.Errorf("failed decoding of ShadowPod: %w", err))
+	}
+
+	// Check existence and get shadow pod origin Cluster ID label
+	clusterID, found := shadowpod.Labels[forge.LiqoOriginClusterIDKey]
+	if !found {
+		webhooklog.Info("Missing origin Cluster ID label")
+		return admission.Denied("missing origin Cluster ID label")
+	}
+
+	code, err := spv.validateShadowPodClusterID(ctx, shadowpod.GetNamespace(), clusterID)
+	if err != nil {
+		if code == http.StatusBadRequest {
+			return admission.Errored(code, err)
+		}
+		return admission.Denied(err.Error())
+	}
+
+	if !spv.enableResourceValidation {
+		return admission.Allowed("")
+	}
+
+	shadowpodlog.Info(fmt.Sprintf("ShadowPod %s decoded: UID: %s - clusterID %s", shadowpod.Name, shadowpod.GetUID(), clusterID))
+
+	// Check existence and get resource offer by Cluster ID label
+
+	resourceoffer, err := liqogetters.GetResourceOfferByLabel(ctx, spv.client, corev1.NamespaceAll,
+		liqolabels.LocalLabelSelector(clusterID))
+	if err != nil {
+		newErr := fmt.Errorf("error getting resource offer by label: %w", err)
+		webhooklog.Info(newErr.Error())
+		return admission.Errored(http.StatusInternalServerError, newErr)
+	}
+
+	clusterName := retrieveClusterName(ctx, spv.client, clusterID)
+
+	webhooklog.Info(fmt.Sprintf("ResourceOffer found for cluster %q with %s ",
+		clusterName, quotaFormatter(resourceoffer.Spec.ResourceQuota.Hard, "Quota")))
+
+	peeringInfo := getOrCreatePeeringInfo(spv.peeringCache, clusterID,
+		clusterName, resourceoffer.Spec.ResourceQuota.Hard)
+
+	spd, err := getOrCreateShadowPodDescription(ctx, spv.client, peeringInfo, shadowpod)
+	if err != nil {
+		shadowpodlog.Info(err.Error())
+		return admission.Denied(err.Error())
+	}
+
+	shadowpodlog.Info(quotaFormatter(spd.quota, "ShadowPod resource limits"))
+
+	err = peeringInfo.testAndUpdateCreation(*spd, *req.DryRun)
+	if err != nil {
+		webhooklog.Info(err.Error())
+		return admission.Denied(err.Error())
+	}
```

Comment:
how do you suggest to solve this? the first way that comes in my mind is to wrap the `getOrCreateShadowPodDescription` inside the `testAndUpdateCreation`. to avoid race conditions i think we need to do the two operation atomically in one step

Reply:
Yes, you should perform them atomically without releasing the lock.